### PR TITLE
Ren ledger

### PR DESCRIPTION
**Pull request by tok-kkk**
**Description**

Ren ledger (formerly know as Hyperdrive) is an Ethereum smart contract which stores all the orders and their priority. Traders needs to open the order on the Ren Ledger first, spending REN to do so, before their orders will be accepted by Darknodes. Darknodes should look up orders in the Ren ledger and match them by their priority.  Both the traders need to send signed message in order to open, and cancel orders, and Darknodes must send signed message to confirm orders.

**Motivation**

We want to have priority for each order so that we can provide prioritisation and so that we can know which orders should be matched, minimising conflicts. We also need a consensus mechanism for resolving conflicts that do occur.

**Design**

The main component will the `RenLedger` contract which will include following fields and functions. 

#### Fields

-  `bytes32[] public orderbook`
  A long list storing all the orders in order of the time received. Each darknode should have its own header from where they should matching the orders.

- `mapping(bytes32 => uint256) public orderPriorities`
  orderMatches maps the orderHash with its priority.

- `mapping(bytes32 => uint8) public orderStatus`
  orderStatus maps the orderHash with its current status. The status can be 0(unknown), 1(open), 2(canceled), 3(confirmed).

- `mapping(bytes32 => bytes32[]) public orderMatches`
  orderMatches maps the orderHash with the counter orders.

- `mapping(bytes32 => Order) public orders`
  orders maps the orderHash with its details.

- `mapping(bytes32 => address) public*orderOwners`
  orderOwners maps the orderHash to the owner address.

**Functions**

- `penOrder (orderID, traderSignature)  public`
  The traderSignature will be the signature of the message "Republic Protocol: open: {orderHash}" signed by the trader

- `confirmOrder (orderID, darknodeSignature)  public`
  The darknodeSignature will be the signature of the message "Republic Protocol: match: {orderHash}" signed by the darknode.
We need to verify the darknode is registered.

- `cancelOrder (orderID, traderSignature)  public`
  The traderSignature will be the signature of the message "Republic Protocol: cancel: {orderHash}" signed by the trader. The trader should be the same with the one who opens the order.

Packages directly modified:
- `republic-sol`

Packages indirectly modified:
- `darknode`
- `relay`

**Unresolved Issues**

# Todo

### CODE REVIEW

**Comment by loongy** [comment]
@tok-kkk 

`mapping(bytes32 => uint8) public orderStatus` should not use a uint8. Instead, we should define a proper enum type for the status. Doing so makes for more meaningful code.

File: contracts/RenLedger.sol [hunk 1]
```diff
@@ -0,0 +1,57 @@
+pragma solidity ^0.4.0;
+
+contract RenLedger {
+
+    enum Parity {Sell, Buy}
+
+    struct Order{
+        uint8 orderType;
```
Comment (loongy) [inline contracts/RenLedger.sol]: Let's introduce an `OrderType` enum for this.

File: contracts/RenLedger.sol [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        Parity  parity;
```
Comment (loongy) [inline contracts/RenLedger.sol]: For consistency, the type should be called `OrderParity`.

File: contracts/RenLedger.sol [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        uint256 expiry;
+    }
+
+    bytes32[] public orderbook;
+
+    mapping(bytes32 => uint256) public orderPriorities;
+
+    mapping(bytes32 => uint8) public orderStatus;
```
Comment (loongy) [inline contracts/RenLedger.sol]: Let's introduce an `OrderStatus` enum for this.

File: contracts/RenLedger.sol [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    mapping(bytes32 => Order) public orders;
+
+    mapping(bytes32 => address) public orderOwners;
+
+    function RenLedger(){
+        orderbook = new bytes32[](0);
+    }
+
+    function OpenOrder(bytes32 orderID, uint8 v, bytes32 r, bytes32 s) public {
```
Comment (loongy) [inline contracts/RenLedger.sol]: You're too used to Golang ;) function names should be `openOrder`.

File: contracts/RenLedger.sol [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        orderStatus[orderID] = uint8(2);
+        orderStatus[counterOrder] = uint8(2);
+    }
+
+    function CancelOrder(bytes32 orderID, uint8 v, bytes32 r, bytes32 s) public{
+        require(ecrecover(orderID, v, r, s) == msg.sender);
+        require(orderStatus[orderID] == uint8(1));
+        require(orderOwners[orderID] == msg.sender);
+
+        orderStatus[orderID] = uint8(3);
```
Comment (loongy) [inline contracts/RenLedger.sol]: This is a great example of why enums are preferred. Using `uint8(3)` has no meaning compared to `OrderStatus.Canceled`.

File: test/unit/RenLedger.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+    let ledger;
+
+    before(async function () {
+        ledger = await renLedger.new();
+    });
+
+    it('should be able to open an order', async function () {
+        for (i = 1 ; i < accounts.length; i ++){
+            let signature = accounts[0].sign(accounts[0].address, 'Republic Protocol: open: 123456');
```
Comment (loongy) [inline test/unit/RenLedger.js]: Should this be `accounts[i]`?

File: contracts/RenLedger.sol [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    modifier onlyDarknode(address _sender) {
+        require(darknodeRegistry.isRegistered(bytes20(_sender)));
+        _;
+    }
+
+    constructor(uint256 _fee, address _republicTokenAddress, address _darknodeRegistryAddress) public {
+        fee = _fee;
+        darknodeRegistry = DarknodeRegistry(_darknodeRegistryAddress);
+        ren = RepublicToken(_republicTokenAddress);
+    }
```
Comment (susruth) [inline contracts/RenLedger.sol]: Please maintain the order of parameters.

File: contracts/RenLedger.sol [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    uint256 public fee;
+    DarknodeRegistry private darknodeRegistry;
+    RepublicToken private ren;
+
+    modifier onlyDarknode(address _sender) {
+        require(darknodeRegistry.isRegistered(bytes20(_sender)));
+        _;
+    }
+
```
Comment (susruth) [inline contracts/RenLedger.sol]: Add comments for all the functions.

**Pull request merged**